rotkehlchen/balances/historical_balances.py

```diff
@@ -0,0 +1,300 @@
+                    usd_value=final_usd,
+                ))
+
+        return historical_balances
+
+    def get_asset_balance(self, asset: 'Asset', timestamp: Timestamp) -> Balance:
+        """Retrieve the historical balance of an asset at a given timestamp.
+
+        Returns the cumulative sum of all non-zero balances up to the given timestamp.
+        Amount and USD value are tracked separately as they may be zero independently.
+        """
+        if not (daily_values := self._get_daily_values(
+            query="""
+            SELECT timestamp, amount, usd_value
+            FROM historical_balance_per_asset
+            WHERE asset_identifier = ? AND timestamp <= ?
+            """,
+            params=(asset.identifier, timestamp),
+        )):
+            return Balance()
+
+        cumulative_amount, cumulative_usd = self._process_cumulative_values(daily_values)
+        return Balance(amount=cumulative_amount, usd_value=cumulative_usd)
+
+    def recreate_historical_balances(self) -> None:
+        """Recreates asset balances from historical events since the last stored point.
+
+        Gets events from DB that resulted in balance changes and processes them to store.
+        """
+        db_history_events = DBHistoryEvents(self.db)
+        with self.db.conn.read_ctx() as cursor:
+            cursor.execute('SELECT MAX(timestamp) FROM historical_balance_per_asset;')
+            events = db_history_events.get_history_events(
+                cursor=cursor,
+                filter_query=HistoryEventFilterQuery.make(
+                    from_ts=Timestamp(cursor.fetchone()[0]),
+                ),
+                has_premium=True,
+            )
+
+            balances: dict[Timestamp, dict[Asset, Balance]] = defaultdict(
+                lambda: defaultdict(lambda: Balance()),  # noqa: PLW0108
+            )
+            for event in events:
+                if (direction := event.maybe_get_direction()) is None or direction == EventDirection.NEUTRAL:  # noqa: E501
+                    log.debug(
+                        'Skipping event for balance processing',
+                        event=event,
+                        reason='no balance effect',
+                    )
+                    continue
+
+                day_timestamp = timestamp_to_daystart_timestamp(ts_ms_to_sec(event.timestamp))
+                if direction == EventDirection.IN:
+                    balances[day_timestamp][event.asset] += event.balance
+                else:  # OUT
+                    balances[day_timestamp][event.asset] -= event.balance
+
+            bindings = []
+            for timestamp, asset_balances in balances.items():
+                for asset, balance in asset_balances.items():
+                    if balance.amount == ZERO:
+                        continue
+
+                    if balance.usd_value == ZERO:
+                        try:
+                            usd_price = PriceHistorian().query_historical_price(
```

Comment:
no no no. This should absolutely not happen. You are now turning a DB-local-only operation to one that will be querying historical prices for all assets, for all days. This will never finish. Price should not go in here.